Third party lib clarification (#1918)
mozilla/extension-workshop: src/content/documentation/publish/third-party-library-usage.md

# Third Party Library Usage

The [Development Practices](/documentation/publish/add-on-policies/#development-practices) section of the [Add-ons Policies](/documentation/publish/add-on-policies/) states that you can include third party libraries in your extension as long the following conditions are met:

* Extensions may only use the release versions of third party libraries.

> * Third party libraries may not be modified.

Reviewer: This is still duplicating content from the policies page. We should avoid that because we can't re-check all related pages when the policies are updated to find copies of provisions that were updated. Wouldn't a link to the policies section be enough, as you do in line 18?

Author: No, I don't think so. This PR was created in response to a request from @mixedpuppy to quote the policy text on this page. The initial motivation for that request was that this page did not communicate the policy requirements to the reader despite the positioning that it is authoritative on the topic.

The other approach I was considering was to move the relevant policy text out of `addons-policy.md` and into a [partial](https://liquidjs.com/tutorials/options.html#Partials-Layouts). That partial would then be imported in both `addons-policy.md` and `third-party-library-usage.md`. But since we don't have an existing pattern of using partials like this I held off.

Reviewer: As I said in private, this is a best practices document, not a policies or requirements document. We should not duplicate official requirements into best practices documents. We can link to the policies where the requirements are stated, but the duplication is a recipe for failure, and we have to avoid that for content that touches policy provisions. There needs to be a single source of truth that we can we can reasonably keep up to date.

# Third Party Library Usage

> > Only release versions of third-party libraries and/or frameworks may be included with an add-on. Modifications to these libraries/frameworks are not permitted.

Reviewer 2: ```suggestion
> Only release versions of third-party libraries and frameworks can be included in an add-on. Modifications to included libraries and frameworks are not permitted.
```

Reviewer: Instead of duplicating the sentence, can we link to the policies section like:

`Please note the requirements for third-party libraries in the #Add-on Policies#`, where `#Add-on Policies`links to the development section of the policies.

Author: Sounds good. I'll update the PR to integrate this feedback after I merge @rebloor's other feedback.

>

> > [Development Practices](/documentation/publish/add-on-policies/#development-practices), [Add-ons Policies](/documentation/publish/add-on-policies/)

Reviewer 2: ```suggestion
> See [Development Practices](/documentation/publish/add-on-policies/#development-practices) and [Add-ons Policies](/documentation/publish/add-on-policies/) for more information.
```

> To complete the review process at [addons.mozilla.org](https://addons.mozilla.org) (AMO), reviewers must be able to verify the code in your extension. If you include third-party libraries in your extension, you will need to provide links to the library source code as part of the AMO submission process.

Reviewer 2: ```suggestion
To complete the [addons.mozilla.org](https://addons.mozilla.org) (AMO) review process, reviewers must be able to verify the code in your extension. If you include third-party libraries in your extension, you must provide links to the library source code as part of the AMO submission process.
```

## When must links for third-party libraries be provided?

Links to third party libraries must be provided when submitting the extension. The links may either be included in the extension's archive, the source archive, or in the "notes for reviewers" on the submission.

> If you include links in one of the archives, we recommend either adding them to a README file in the root directory or another root file clearly intended for AMO reviewers.

Reviewer: While there are many options, in a best practices document let's aim to pick the best practice. The challenge here is that there is no one-size-fits all. For example, if there is no other reason to attach source code, we don't want the developer to do it just for third-party libraries.

In general, notes to reviewers is fine, unless the amount of libraries (and other reviewer information) is too high to be reasonable for a "Notes" field.

Author: Makes sense. In that case I think we can simplify this text a bit. WDYT?

```suggestion
Links to third party libraries must be provided when submitting a new version to AMO. You can add the links to the "Notes for Reviewers" section of your extension's details.
```

Reviewer 2: ```suggestion
When submitting a version to AMO, links to third-party libraries must be provided. You can add the links to the "Notes for Reviewers" section of your extension's details. 
```

Reviewer: Either is fine with me, thanks.

{% capture content %}

## Use of package managers

Extensions developers can use package managers and package repositories like [npm](https://www.npmjs.com/) to retrieve third party libraries.

With a default npm configuration, third party library dependencies are declared in the project's `package.json` file: this qualifies as a third party library link as [previously described](#how-to-determine-the-third-party-library-link).

> Reviewers must be be able to retrieve and review all packages used by your extension, so the use of private packages or non-public registries is not recommended. If you do, you will need to provide reviewers with access to the private packages. Access information should be provided along with build instructions as part of the [source code submission](https://extensionworkshop.com/documentation/publish/source-code-submission/) process.

Reviewer: ```suggestion
Reviewers must be be able to retrieve and review all packages used by your extension, so the use of private packages or non-public registries is not recommended. If you do, include those as "pre-installed" dependencies in your `node-modules` folder. Access information should be provided along with build instructions as part of the [source code submission](https://extensionworkshop.com/documentation/publish/source-code-submission/) process.
```

I am not sure if "pre-installed" is the right term here or descriptive enough. Thoughts on that wording?

Author: I don't think "pre-installed" will quite get the idea across. I took another shot at making this passage more explicit. WDYT?

```suggestion
Reviewers must be able to retrieve and review all packages used by your extension. As a result, the use of private packages or non-public registries is permissible but not recommended. If you use non-public dependencies, you will either need to include the relevant `node_modules` sub-directories in your [source code submission](/documentation/publish/source-code-submission/) or include access credentials and instructions in your build instructions.
```

Reviewer 2: ```suggestion
Reviewers must be able to retrieve and review all packages used by your extension. Therefore, the use of private packages or non-public registries is permissible but not recommended. If you use non-public dependencies, you must include the relevant `node_modules` sub-directories in your [source code submission](/documentation/publish/source-code-submission/) or access credentials and instructions in your build instructions.
```

Reviewer: Either works for me, but let's remove the `or include access credentials and instructions in your build instructions.` part. We will need to consider and evaluate that first. Some developers have done it in the past by themselves and anecdotally, that lead to issues in reproducing the build.
